lib/hsa/activity_prof.h

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright (c) 2015-2016 Advanced Micro Devices, Inc. All rights reserved.
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANNTY OF ANY KIND, EXPRESS OR
+IMPLIED, INNCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANNY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER INN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR INN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#ifndef ACTIVITY_PROF_H
+#define ACTIVITY_PROF_H
+
+#include <mutex>
+
+#include "kalmar_runtime.h"
+
+#if USE_PROF_API
+#include "prof_protocol.h"
+#include "hsa_rt_utils.hpp"
+
+#define ACTIVITY_PROF_INSTANCES()                                                    \
+    namespace activity_prof {                                                        \
+        CallbacksTable::table_t CallbacksTable::_table{};                            \
+        CallbacksTable::mutex_t CallbacksTable::_mutex;                              \
+        template<> HSAOp::ActivityProf::timer_t* HSAOp::ActivityProf::_timer = NULL; \
+        template<> std::atomic<record_id_t> HSAOp::ActivityProf::_glob_record_id(0); \
+    } // activity_prof
```

Comment:
Could you change this to avoid using a macro?  It makes the code that uses this macro (`mcwamp_hsa.cpp`) looks very confusing (a function call at the file scope) and it really hurts readability

Reply:
The macro is used to instantiate static fields.
I tried to incapsulate all implementation in the header.
I see two solutions:
a) Actually the header is included only in one file mcwamp_hsa.cpp so I can just put the instantiation directly in the header.

b) Also I can do the instantiation explicitly in  mcwamp_hsa.cpp under USE_PROF_API macro.

Which solution do you prefer or maybe you have another proposal?

Reply:
b) seems fine

Reply:
a) is fine